haystack/components/rankers/sentence_transformers_diversity.py

```diff
@@ -218,17 +234,88 @@ def _greedy_diversity_order(self, query: str, documents: List[Document]) -> List
 
         return ranked_docs
 
+    def _embed_and_normalize(self, query, texts_to_embed):
+        # Calculate embeddings
+        doc_embeddings = self.model.encode(texts_to_embed, convert_to_tensor=True)  # type: ignore[attr-defined]
+        query_embedding = self.model.encode([self.query_prefix + query + self.query_suffix], convert_to_tensor=True)  # type: ignore[attr-defined]
+
+        # Normalize embeddings to unit length for computing cosine similarity
+        if self.similarity == "cosine":
+            doc_embeddings /= torch.norm(doc_embeddings, p=2, dim=-1).unsqueeze(-1)
+            query_embedding /= torch.norm(query_embedding, p=2, dim=-1).unsqueeze(-1)
+        return doc_embeddings, query_embedding
+
+    def _maximum_margin_relevance(
+        self, query: str, documents: List[Document], lambda_threshold: float = 0.5
+    ) -> List[Document]:
+        """
+        Orders the given list of documents according to the Maximum Margin Relevance (MMR) scores.
+
+        MMR scores are calculated for each document based on their relevance to the query and diversity from already
+        selected documents.
+
+        The algorithm iteratively selects documents based on their MMR scores, balancing between relevance to the query
+        and diversity from already selected documents. The 'lambda_threshold' controls the trade-off between relevance
+        and diversity.
+
+        A closer value to 0, favors diversity, while a closer value to 1, favors relevance to the query.
+
+        See : "The Use of MMR, Diversity-Based Reranking for Reordering Documents and Producing Summaries"
+               https://www.cs.cmu.edu/~jgc/publication/The_Use_MMR_Diversity_Based_LTMIR_1998.pdf
+        """
+
+        texts_to_embed = self._prepare_texts_to_embed(documents)
+
+        doc_embeddings, query_embedding = self._embed_and_normalize(query, texts_to_embed)
+
+        n = len(documents)
+        selected: List[int] = []
+
+        # compute MMR score for each document
```

Comment:
I see some potential ways to improve this algorithm:
1. keep track of the best documents, to avoid reranking
2. consider `top_k` for selection

Reply:
There's already a global `top_k` used at run time, I don't think it makes sense to use another one here - same philosophy as in `greedy_diversity()`

> 1. keep track of the best documents, to avoid reranking

What do you you mean by best documents?

Reply:
I think that looking at [a different implementation of MMR](https://api.python.langchain.com/en/latest/_modules/langchain_astradb/utils/mmr.html#maximal_marginal_relevance) might clarify my comments.